### PR TITLE
op-node: Add Epoch Hash to batch

### PR DESCRIPTION
**Pull request by trianglesphere**
**Description**
This commits a batch to a specific L1 origin block by hash rather
than just by number. This help in the case of L1 reorgs by stopping
batches from being applied in weird ways.

**Additional context**
This does not add a block number to the batch. The plan for eager batch derivation
is to continue to use timestamps, but slightly modify derivation rules (similar to how
they would be modified even if it was switched over to using block numbers).

**Metadata**
- Fixes #[Link to Issue]

### CODE REVIEW

File: op-node/rollup/derive/batches.go [hunk 1]
```diff
@@ -34,12 +35,15 @@ func FilterBatches(log log.Logger, config *rollup.Config, epoch rollup.Epoch, mi
 	return
 }
 
-func ValidBatch(batch *BatchData, config *rollup.Config, epoch rollup.Epoch, minL2Time uint64, maxL2Time uint64) error {
+func ValidBatch(batch *BatchData, config *rollup.Config, epoch rollup.Epoch, epochHash common.Hash, minL2Time uint64, maxL2Time uint64) error {
```
Comment (protolambda) [inline op-node/rollup/derive/batches.go:38]: Maybe we should just pass `epoch eth.BlockID`?

File: op-node/rollup/derive/batches.go [hunk 2]
```diff
@@ -5,16 +5,17 @@ import (
...
 	"github.com/ethereum-optimism/optimism/op-node/rollup"
+	"github.com/ethereum/go-ethereum/common"
 	"github.com/ethereum/go-ethereum/core/types"
 	"github.com/ethereum/go-ethereum/log"
 )
 
 var DifferentEpoch = errors.New("batch is of different epoch")
 
-func FilterBatches(log log.Logger, config *rollup.Config, epoch rollup.Epoch, minL2Time uint64, maxL2Time uint64, batches []*BatchData) (out []*BatchData) {
+func FilterBatches(log log.Logger, config *rollup.Config, epoch rollup.Epoch, epochHash common.Hash, minL2Time uint64, maxL2Time uint64, batches []*BatchData) (out []*BatchData) {
```
Comment (protolambda) [inline op-node/rollup/derive/batches.go:15]: Same here

File: op-node/rollup/derive/batches.go [hunk 3]
```diff
@@ -62,7 +66,7 @@ func ValidBatch(batch *BatchData, config *rollup.Config, epoch rollup.Epoch, min
 }
 
 // FillMissingBatches turns a collection of batches to the input batches for a series of blocks
-func FillMissingBatches(batches []*BatchData, epoch, blockTime, minL2Time, nextL1Time uint64) []*BatchData {
+func FillMissingBatches(batches []*BatchData, epochNumber uint64, epochHash common.Hash, blockTime, minL2Time, nextL1Time uint64) []*BatchData {
```
Comment (protolambda) [inline op-node/rollup/derive/batches.go:69]: Same here

File: op-node/rollup/derive/batches.go [hunk 4]
```diff
@@ -34,12 +35,15 @@ func FilterBatches(log log.Logger, config *rollup.Config, epoch rollup.Epoch, mi
...
 }
 
-func ValidBatch(batch *BatchData, config *rollup.Config, epoch rollup.Epoch, minL2Time uint64, maxL2Time uint64) error {
+func ValidBatch(batch *BatchData, config *rollup.Config, epoch rollup.Epoch, epochHash common.Hash, minL2Time uint64, maxL2Time uint64) error {
 	if batch.Epoch != epoch {
 		// Batch was tagged for past or future epoch,
 		// i.e. it was included too late or depends on the given L1 block to be processed first.
 		return DifferentEpoch
 	}
+	if batch.OriginHash != epochHash {
```
Comment (protolambda) [inline op-node/rollup/derive/batches.go:44]: Can merge this with above check, and return a wrapped error around `DifferentEpoch` with the two different id values (batch vs epoch).

**Review by protolambda — COMMENTED** [review]
Generally looks good, maybe consider using `eth.BlockID` (and we'd get rid of the `rollup.Epoch` type then I guess).

**Comment by trianglesphere** [comment]
switched to `eth.BlockID` in calling code

File: op-node/rollup/derive/batches.go (hunk 4 above)
Comment (trianglesphere) [reply op-node/rollup/derive/batches.go:44]: check merged, but no differentiation in the error case

File: op-node/rollup/derive/channel_out.go [hunk 5]
```diff
@@ -164,7 +164,7 @@ func blockToBatch(block *types.Block, w io.Writer) error {
 	}
 
 	batch := &BatchData{BatchV1{
-		Epoch:        rollup.Epoch(l1Info.Number),
+		EpochNum:     rollup.Epoch(l1Info.Number),
```
Comment (protolambda) [inline op-node/rollup/derive/channel_out.go:167]: think you missed the epoch hash here, I'll add it

**Pull request merged**